### DISCUSSION TITLE
How to `git reset --soft HEAD~1` ?

### DISCUSSION THREAD

**Discussion by dgr** [magit/magit#5215] [Support] [answered]
I'm trying to use Magit to `git reset --soft HEAD~1` and it doesn't seem to work. Specifically, after hitting `X` and `s`, when I then type `HEAD~1` at the "Soft reset master to:" prompt, it doesn't like it and gives me an error about flyspell (I have no idea why). Can I not specify a commit using `HEAD~1` syntax? It also doesn't seem to accept a SHA (whether partial or complete). It gives the same error message about flyspell. Bug? This is with the latest (20240831.2255) off MELPA.

**Comment by dgr** [comment]
Forgot to say that I'm also running Ivy. Perhaps that's interfering?

**Comment by kyleam** [comment]
> Can I not specify a commit using `HEAD~1` syntax? It also doesn't seem to accept a SHA (whether partial or complete).

Each of those is valid input.  I don't use Ivy on a regular basis, but I've verified that `magit-reset-soft` works as expected on my end (in particular I can reset to `HEAD~1` or an arbitrary commit).  This was with the latest Magit (0aa26864e) and Ivy (Swiper 8dc02d5b7) and a vanilla configuration aside from

```
(ivy-mode 1)
```

A good first step might be checking if you still see the issue with a minimal config.

**Reply by dgr** [reply]
OK, thanks for confirming that those all should work (I was shocked when they didn't in my case). I'll try to strip back my Emacs config and figure it out. If I find anything other than pilot error, I'll report back.

**Reply by dgr** [reply]
Yep, works fine with a minimal config. Sorry to have troubled you. Something else is conflicting.

**Reply by dgr** [reply]
I isolated it down to flyspell. Ivy didn't seem to have anything to do with it. This is my `use-package` in `~/.emacs.d/init.el`.
```emacs lisp
(use-package flyspell
  :ensure t
  :hook text-mode
  :config
  (setq ispell-program-name "aspell"
        ispell-list-command "--list"))
```

If I don't hook text-mode, then everything works with Magit. Any thoughts on that? That's about as minimal of a config for flyspell as I can think of.

**Reply by kyleam** [reply]
> If I don't hook text-mode, then everything works with Magit. Any thoughts on that?

Interesting.  I don't have any ideas what could be going on there.  `minibuffer-mode` doesn't inherit from `text-mode`, so I don't how adding flyspell-model  would come into play.

Fwiw I tried to reproduce the issue with the config below but wasn't able to.

```
(require 'magit)
(require 'flyspell)
(add-hook 'text-mode-hook #'flyspell-mode)
```

That's with an Emacs built from a recent commit (2ca7d5649c6) on the emacs-30 branch.

What's the flyspell error you're seeing?

**Reply by dgr** [reply, marked as answer]
The error is `.../flyspell.elc failed to define function flyspell`. So, it's trying to load flyspell and failing because it has the wrong function name (should be `flyspell-mode`).

Works:

```emacs lisp
(require 'package)
(package-initialize)

(use-package flyspell
  :ensure t
  :hook (text-mode . flyspell-mode) ; specify flyspell-mode explicitly
  :config
  (setq ispell-program-name "aspell"
        ispell-list-command "--list"))

(use-package magit
  :ensure t
  :bind (("C-x g" . magit-status))
  :config
  (setq magit-git-executable "/opt/homebrew/bin/git")
  (setq magit-auto-revert-immediately t))
```

Doesn't work:
```emacs lisp
(require 'package)
(package-initialize)

(use-package flyspell
  :ensure t
  :hook text-mode ; let use-package figure out the mode name, which should be flyspell-mode but is flyspell instead
  :config
  (setq ispell-program-name "aspell"
        ispell-list-command "--list"))

(use-package magit
  :ensure t
  :bind (("C-x g" . magit-status))
  :config
  (setq magit-git-executable "/opt/homebrew/bin/git")
  (setq magit-auto-revert-immediately t))
```

`use-package` is supposed to add a hook to `xyz-mode` when you just use `:hook abc-mode` within a `(use-package xyz ...)`. Seems like either a bug or a change in behavior for `use-package`. For whatever reason, it's adding a hook to `flyspell` and not `flyspell-mode`.

Regardless, this is not a Magit bug or problem. Sorry to have troubled you. Thanks for all your help.

**Reply by dgr** [reply]
All that said, I'm not sure why Emacs was trying to turn on text-mode and firing that hook when Magit was doing a soft reset. That was a bit odd.
